.github/workflows/cd.yaml

```diff
@@ -109,36 +112,123 @@ jobs:
+        with:
+          fetch-depth: 0
+          token: ${{ secrets.RELEASE_GITHUB_TOKEN }}
+
+      - name: Install poetry
+        run: pipx install poetry
+
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+          cache: "poetry"
+
+      - name: Install dependencies
+        run: poetry install --no-interaction --with cicd
+
       - name: Publish to GitHub
-        if: steps.get_branch.outputs.branch == 'main' && !inputs.dry_run && inputs.publish_github
         run: poetry run semantic-release $DRY_RUN publish
         env:
             GITHUB_TOKEN: ${{ secrets.RELEASE_GITHUB_TOKEN }}
 
+  append-binaries-to-release:
+    name: Append Binaries to GitHub Release
+    needs: [prepare-release, publish-to-github, build-and-upload-binaries]
+    if: needs.prepare-release.outputs.branch_name == 'main' && !inputs.dry_run && inputs.publish_github
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    steps:
+      - name: Download all artifacts
+        uses: actions/download-artifact@v4
+        with:
+          path: all-artifacts
+          pattern: binary-artifacts-*
+          merge-multiple: true
+
+      - name: Display structure of downloaded files
+        run: find all-artifacts -type f | sort
+          
+      - name: Append artifacts to release
+        uses: softprops/action-gh-release@v1
+        with:
+          fail_on_unmatched_files: true
+          files: all-artifacts/**/*
+          tag_name: ${{ format('v{0}', needs.prepare-release.outputs.version_number) }}
+
+  publish-to-pypi:
+    name: Publish to PyPI
+    needs: [prepare-release, append-binaries-to-release]
+    if: needs.prepare-release.outputs.branch_name == 'main' && !inputs.dry_run && inputs.publish_pypi
+    runs-on: ubuntu-latest
+    permissions:
+      id-token: write
+      contents: read
+    steps:
+      - name: Download release artifacts
+        uses: actions/download-artifact@v4
+        with:
+          name: release-artifacts
+          path: release-artifacts
+
       - name: Publish to PyPI - Compiler
-        if: steps.get_branch.outputs.branch == 'main' && !inputs.dry_run && inputs.publish_pypi && inputs.compiler_wheel
+        if: inputs.compiler_wheel
         uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          packages-dir: release-artifacts/dist
 
       - name: Publish to PyPI - Stubs
-        if: steps.get_branch.outputs.branch == 'main' && !inputs.dry_run && inputs.publish_pypi && inputs.stubs_wheel
+        if: inputs.stubs_wheel
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
-          packages-dir: stubs/dist
+          packages-dir: release-artifacts/stubs/dist
 
   publish-docs:
     name: Publish Docs
-    if: ${{ !inputs.dry_run && inputs.publish_docs }}
+    needs: [prepare-release]
+    if: needs.prepare-release.outputs.branch_name == 'main' && !inputs.dry_run && inputs.publish_docs
```

Comment:
@daniel-makerx @achidlow do we want to keep it as is or given new sequential flow -> run it only after all of the previous steps are succeeded ? Kept it as is given that in case of a failed release documentation can still be reverted via separate workflow trigger